File: week-7/assignment-7/java-ride-sharing-system/src/com/rideshare/Worker.java
package com.rideshare;

import java.util.List;

/**
 * Worker thread.
 * - Processes tasks, simulates delay, and writes results.
 * - Manages concurrency and resource access.
 * - Exception handling and logging per thread.
 */
public class Worker implements Runnable {
    private final int id;
    private final TaskQueue taskQueue;
    private final List<String> results;

    public Worker(int id, TaskQueue taskQueue, List<String> results) {
        this.id = id;
        this.taskQueue = taskQueue;
        this.results = results;
    }

    @Override
    public void run() {
        LoggerConfig.LOGGER.info("Worker " + id + " started.");
        try {
            String task;
            while ((task = taskQueue.getTask()) != null) {
                LoggerConfig.LOGGER.info("Worker " + id + " processing: " + task);
                try {
                    // Simulate processing delay (computational work)
                    Thread.sleep(500);
                } catch (InterruptedException e) {
                    // Handle InterruptedException
                    LoggerConfig.LOGGER.warning("Worker " + id + " interrupted.");
                    Thread.currentThread().interrupt();
                    break;
                }
                // Synchronize results list access to prevent race conditions
                synchronized (results) {
                    results.add("Completed by Worker " + id + ": " + task);
                }
            }
        } catch (Exception e) {
            // Handle any unexpected exceptions
            LoggerConfig.LOGGER.severe("Worker " + id + " encountered error: " + e.getMessage());
        } finally {
            LoggerConfig.LOGGER.info("Worker " + id + " finished.");
        }
    }
}

File: week-7/assignment-7/go-ride-sharing-system/pkg/worker/worker.go
package worker

import (
	"fmt"
	"sync"
	"time"

	"rideshare/pkg/logger"
	"rideshare/pkg/taskqueue"
)

/*
	Worker goroutines processing tasks
	Safe concurrent access via channels
	Error handling with defer and recover
*/

// Worker processes tasks from the queue
func Worker(id int, taskQueue *taskqueue.TaskQueue, results *[]string, mu *sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()

	logger.Logger.Println(fmt.Sprintf("Worker %d started.", id))

	for task := range taskQueue.Tasks {
		logger.Logger.Println(fmt.Sprintf("Worker %d processing: %s", id, task))

		// Simulate processing delay (computational work)
		time.Sleep(500 * time.Millisecond)

		// Safely append result (sync on results slice)
		mu.Lock()
		*results = append(*results, fmt.Sprintf("Completed by Worker %d: %s", id, task))
		mu.Unlock()
	}

	logger.Logger.Println(fmt.Sprintf("Worker %d finished.", id))
}
